README
Author:  Andrea Fabrizi
Email:   andrea.fabrizi@gmail.com
Web:     http://www.andreafabrizi.it
Twitter: @andreaf83

Dropbox Uploader is a BASH script which can be used to upload, download, delete or list files from Dropbox, an online file sharing, synchronization and backup service. 

It's written in BASH scripting language and only needs cURL.

Why use this script?

Portable: It's written in BASH scripting language and only needs cURL (curl is a tool to transfer data from or to a server, available for all operating systems and installed by default in many linux distributions).
Secure:It's not required to provide your username/password to this script, because it uses the official Dropbox API for authentication process. 

USAGE:
======

Usage: ./dropbox_uploader.sh COMMAND [PARAMETERS]...

[%%]: Required param 
<%%>: Optional param

Commands:

 upload   [LOCAL_FILE]  <REMOTE_FILE> 
 - Upload local file to remote Dropbox folder
   If the file is bigger than 150Mb the file is uploaded using small chunks (default 4Mb); in this case, if VERBOSE is set to 1, a . (dot) is printed for every chunk successfully uploaded. Instead, if an error occurs during the chunk uploading, an * (star) is printed and the upload is retried for a maximum of three times.
   Only if the file is smaller than 150Mb, the standard upload API is used, and if VERBOSE is set to 1 the default curl progress bar is displayed during the upload process.

 download [REMOTE_FILE] <LOCAL_FILE>  
 - Download file from Dropbox to local folder

 delete   [REMOTE_FILE/REMOTE_DIRECTORY] 
 - Remove a remote file from Dropbox

 mkdir    [REMOTE_DIRECTORY]
 - Create a remote directory on DropBox

 list     <REMOTE_DIRECTORY>
 - List contents of the remote Dropbox folder

 share [REMOTE_FILE]
 - Get a public share link for the specified file
 
 info          
 - Print some info about your Dropbox account

 unlink
 - Unlink this script from your Dropbox account


Optional parameters:
 -f [FILENAME]
    Load the configuration file from a specific file

Examples:
    ./dropbox_uploader.sh upload /etc/passwd /myfiles/passwd.old
    ./dropbox_uploader.sh upload /etc/passwd
    ./dropbox_uploader.sh download /backup.zip
    ./dropbox_uploader.sh delete /backup.zip
    ./dropbox_uploader.sh mkdir /myDir/
    ./dropbox_uploader.sh upload "My File.txt" "My File 2.txt"   (File name with spaces...)
    ./dropbox_uploader.sh share "My File.txt"


TESTED ENVIRONMENTS:
====================
GNU Linux
FreeBSD 8.3
MacOSX
Windows/Cygwin
Raspberry Pi
QNAP
iOS

If you have successfully tested this script on others systems or platforms please let me know!


HOW TO USE A PROXY:
===================

To use a proxy server, simply set the "https_proxy" environment variable:

### Linux:

    export HTTP_PROXY_USER=XXXX
    export HTTP_PROXY_PASSWORD=YYYY
    export https_proxy=http://192.168.0.1:8080

### BSD: 

    setenv HTTP_PROXY_USER XXXX
    setenv HTTP_PROXY_PASSWORD YYYY
    setenv https_proxy http://192.168.0.1:8080

   
BASH AND CURL INSTALLATION:
===========================

### Debian & Ubuntu Linux: 
    
    sudo apt-get install bash (Probably bash is already installed on your system)
    sudo apt-get install curl

### BSD: 

    cd /usr/ports/shells/bash && make install clean
    cd /usr/ports/ftp/curl && make install clean

### Cygwin:

    You need to install these packages: 
    - curl
    - ca-certificates

### Build cURL from source: 

    Download the source tarball from http://curl.haxx.se/download.html
   

DROPBOX UPLOADER FIRST START:
=============================

 $chmod +x dropbox_uploader.sh
 $./dropbox_uploader.sh


CHANGELOG:
==========

Version 0.11.7 - 23 Apr 2013
- Fixed issue with special chars
- Fix for iOS

Version 0.11.6 - 15 Mar 2013
- Add optional command-line parameter ('-f') to read dropbox configuration from a specific file (thanks to pjv)

Version 0.11.5 - 22 Gen 2013
- Added the ability to get a share link for a specified file (thanks to camspiers)

Version 0.11.4 - 17 Gen 2013
- Fix for QNAP compatibility (thanks to Fritz Ferstl)
- Implemented mkdir command (thanks to Joel Maslak)
- Fix for Solaris compatibility

Version 0.11.3 - 22 Dec 2012:
- Improved list command (thanks to Robert González)
- Fixed problem with unicode characters

Version 0.11.2 - 14 Nov 2012:
- Added a check for the free quota before uploading a file
- Now the quota informations are displayed in Mb
- Removed urlencode function for incompatibility with older curl versions
- Fixed problem uploading files that contains @ character
- Minor changes

Version 0.11.1 - 12 Nov 2012:
- As suggested by the DropBox API documentation, the default chunk for chunked uploads is now 4Mb
- Minor changes

Version 0.11 - 11 Nov 2012:
- Parameterized the curl binary location
- Fix for MacOSX 10.8 (thanks to Ben - www.aquiltforever.com)

Version 0.10 - 03 Nov 2012:
- Code clean
- Improved urlencode function (thanks to Stefan Trauth - www.stefantrauth.de)
- Added command remove as alias of delete
- Fix for Raspberry PI
- Now if an error occurs during a chunk uploading, the upload is retried for a maximum of three times
- Minor changes
- Tested on Cygwin and MacOSX

Version 0.9.9 - 24 Oct 2012:
- Added the possibility to choose the access level (App folder o Full Dropbox) during the setup procedure
- Added a check for the BASH shell version
- Fixed problems in listing files/directories with special characters
- Added the option CURL_ACCEPT_CERTIFICATES (see the script source)
- Added back the standard upload function. Now only if the file is greater than 150Mb, the chunked_upload API will be used.
- Fixed compatibility with bsd sed. Tested on FreeBSD, but probably it works on others bsd versions and osx. Let me know!
- Minor changes

Version 0.9.8 - 03 Oct 2012:
- Implemented chunked upload. Now there is no limit to file size!

Version 0.9.7 - 14 Sep 2012:
- Fixed bug in listing empty directories

Version 0.9.6 - 12 Sep 2012:
- Implemented list command
- Minor changes

Version 0.9.5 - 18 Jul 2012:
- Added a check for the maximum file size allowed by the DropBox API
- Minor changes

Version 0.9.4 - 19 Mar 2012:
- Implemented delete command
- Minor changes

Version 0.9.3 - 01 Mar 2012:
- Implemented download command
- Improved info output
- Fixed utime function
- Added dependency check for basename
- The script always returns 1 when errors occurs
- Improved error handling
- Fixed problem with spaces in config file name
- Minor bug fixes

Version 0.9.2 - 28 Feb 2012:
- Increased security, now any user can create his own Dropbox App

Version 0.9.1 - 27 Feb 2012:
- Fixed problem with spaces in dst file name

Version 0.9 - 27 Feb 2012:
- Code rewritten from scratch (CLI changed)
- Improved security and stability using official dropbox API, no more username/password needed!

Version 0.8.2 - 07 Sep 2011:
- Removed INTERACTIVE_MODE variable (now the progress bar is shown in VERBOSE mode)
- Improved command line interface and error messages
- Minor bug fixes

Version 0.8.1 - 31 Aug 2011 (by Dawid Ferenczy - www.ferenczy.cz)
- added prompt for the Dropbox password from keyboard, if there is no password 
  hardcoded or given as script command line parameter (interactive mode)
- added INTERACTIVE_MODE variable - when set to 1 show CURL progress bar.
  Set to 1 automatically when there is no password hardcoded or given as
  parameter. Controls verbosity of CURL.

Version 0.7.1 - 10 Mar 2011:
- Minor bug fixes

Version 0.7 - 10 Mar 2011:
- New command line interface
- Code clean

Version 0.6 - 11 Gen 2011:
- Fixed issue with spaces in file/forder name

Version 0.5 - 04 Gen 2011:
- Recursive directory upload

Version 0.4 - 29 Dec 2010:
- Now works on BSD and MAC
- Interactive prompt for username and password
- Speeded up the uploading process
- Debug mode

Version 0.3 - 18 Nov 2010:
- Regex updated

Version 0.2 - 04 Sep 2010:
- Removed dependencies from tempfile
- Code clean

Version 0.1 - 23 Aug 2010:
- Initial release


CHANGELOG.md
# CHANGELOG

## Version _._ - __ ___ 20__
* Fix issue #371: Creating directory exits 1 if it already exists

## Version 1.0 - 10 Aug 2016
* Add search option
* Increase chunk size to 50MB for better performance
* Add testUnit script
* Upgrade to API v2
* Fix #278: Documentation issue for command line
* Add -h option to show human readable file sizes
* Fix issue #240: the connection is retried in case of error for chunked uploads
* The file name encoding problems including the issue #139 should be fixed now
* Add system info to debug output
* Fix issue #230: Version String isn't updated in dropbox_uploader
* Fixed issue #227

## Version 0.16 - 31 Jul 2015
* Add saveurl command to download remote files directly into a dropbox folder
* Fixed issue #174: $DU needs quoting in dropShell
* The share command will now return by default the short url
* Improved portability of the file_size function
* Fixed minor bash compatibility issue in the urlencode function
* Improved normalize_path function and minor fix
* Removed deprecated db_free_quota function

## Version 0.15 - 08 Jun 2015
* Fix issue #163, List only show folders but not files
* Fix issue #168, Share call not returning a share link
* Fix issue #172, list command always return an unnamed dir on an empty directory
* Fix issue #165, Issue with uploading a file to a folder
* Fix issue #162, Uploading a folder to a folder that already exists on dropbox has unexpected behaviour
* Updated setup procedure

## Version 0.14 - 15 Jul 2014
* Fix issue #112, replaced bash condition -f with -e
* Fix issue #117, sed: 1: invalid command code z
* Fix issue #122, Directory upload doesn't work
* Fix issue #132, enhanced cross compatibility for bash conditions

## Version 0.13 - 17 Jan 2014
* Minor languages fixes, based on issue #88
* Fix issue #91, sed take long time to process regexp on big data
* Fix issue #87, second chunked upload fails after successful first
* Fix issue #97, greadlink not selected on OSX Mavericks
* Fix issues #99 and #100: The stat command in OSX 10.9 has no parameter named --format
* Fix issue #99, clarified how to install greadlink (via homebrew) for Mac OS X. Thanks to paultreny
* Improved list command and minor changes
* Fix issue #103, transform Unicode code-points to characters in list command
* Add VERSION output in DEBUG mode
* DropShell: Fixed issue with quoted command parameters
* Fix issue with absolute paths in CD command
* Enhancement request #108: add initial wildcard support in upload command
* Now the destination file/dir for the upload command is mandatory
* Fixed issue listing files/folders names containing quotes

## Version 0.12.1 - 24 Oct 2013
* Fixed unicode problem with special chars in file names
* Fixed problem removing temp files
* Improved detection of configuration error and better handling of http errors
* Updated setup procedure
* Fixed incompatibility with old bash version

## Version 0.12 - 06 Sep 2013
* Fixed problem uploading twice the same directory
* Add checking for not allowed file names
* Handling connection/resolving host errors
* Better error handling
* Converted if condition to BASH native syntax
* Add normalize_path function, to canonicalize paths
* Add -s option to skip existing files when upload/download
* Removed check for free quota on every file upload to improve performance
* Add checks for file permissions before writing/reading files on the local fs
* Add copy function
* Fixed unicode problem in file listing
* A lot of minor changes and improvements

## Version 0.11.9 - 27 Jun 2013
* Add missing -f option in usage
* Add some /dev/null redirects
* Improved error messages
* Improved grep syntax
* Add function to check the cURL exit code
* Improved messages output
* Add new command line options: -q -d -k

## Version 0.11.8 - 05 Jun 2013
* Add move/rename function
* Improved the configuration file management (thanks to Robert G.)
* Updated strings to reflect the new Dropbox "Create App" page
* Add support for download directories
* Add support for upload directories

## Version 0.11.7 - 23 Apr 2013
* Fixed issue with special chars
* Fix for iOS

## Version 0.11.6 - 15 Mar 2013
* Add optional command-line parameter ('-f') to read dropbox configuration from a specific file (thanks to pjv)

## Version 0.11.5 - 22 Gen 2013
* Added the ability to get a share link for a specified file (thanks to camspiers)

## Version 0.11.4 - 17 Gen 2013
* Fix for QNAP compatibility (thanks to Fritz Ferstl)
* Implemented mkdir command (thanks to Joel Maslak)
* Fix for Solaris compatibility

## Version 0.11.3 - 22 Dec 2012:
* Improved list command (thanks to Robert González)
* Fixed problem with unicode characters

## Version 0.11.2 - 14 Nov 2012:
* Added a check for the free quota before uploading a file
* Now the quota informations are displayed in Mb
* Removed urlencode function for incompatibility with older curl versions
* Fixed problem uploading files that contains @ character
* Minor changes

## Version 0.11.1 - 12 Nov 2012:
* As suggested by the DropBox API documentation, the default chunk for chunked uploads is now 4Mb
* Minor changes

## Version 0.11 - 11 Nov 2012:
* Parameterized the curl binary location
* Fix for MacOSX 10.8 (thanks to Ben - www.aquiltforever.com)

## Version 0.10 - 03 Nov 2012:
* Code clean
* Improved urlencode function (thanks to Stefan Trauth * www.stefantrauth.de)
* Added command remove as alias of delete
* Fix for Raspberry PI
* Now if an error occurs during a chunk uploading, the upload is retried for a maximum of three times
* Minor changes
* Tested on Cygwin and MacOSX

## Version 0.9.9 - 24 Oct 2012:
* Added the possibility to choose the access level (App folder o Full Dropbox) during the setup procedure
* Added a check for the BASH shell version
* Fixed problems in listing files/directories with special characters
* Added the option CURL_ACCEPT_CERTIFICATES (see the script source)

testUnit.sh
#!/bin/bash

CURRENT_PATH=$(cd `dirname $0`; pwd)
DU=$CURRENT_PATH/dropbox_uploader.sh

function check_exit
{
    if [ $? -ne 0 ]; then
        echo " Error!!!"
        exit 1
    else
        echo " Passed"
    fi
}

#Creating garbage data
echo -ne " - Creating garbage data...\n"
rm -fr "testData"
mkdir -p "testData"
dd if=/dev/urandom of="testData/file 1.txt" bs=1M count=3
dd if=/dev/urandom of="testData/file 2 ù.txt" bs=1M count=5
mkdir -p "testData/recurse"
dd if=/dev/urandom of="testData/recurse/file 3.txt" bs=1M count=1
dd if=/dev/urandom of="testData/recurse/test_Ü.txt" bs=1M count=1
mkdir -p "testData/recurse/dir 1/"
dd if=/dev/urandom of="testData/recurse/dir 1/file 4.txt" bs=1M count=1
mkdir -p "testData/recurse/dir 1/dir 3/"
dd if=/dev/urandom of="testData/recurse/dir 1/dir 3/file 5.txt" bs=1M count=1
mkdir -p "testData/recurse/dir 2/"

rm -fr recurse

#Rmdir
echo -ne " - Remove remote directory..."
$DU -q remove du_tests
echo ""

#Mkdir
echo -ne " - Make remote directory..."
$DU -q mkdir du_tests
check_exit

#Simple upload
echo -ne " - Simple file upload..."
$DU -q upload "testData/file 1.txt" du_tests
check_exit

#Checking with list
echo -ne " - Checking file..."
$DU -q list du_tests | grep "file 1.txt" > /dev/null
check_exit

#Simple upload 2
echo -ne " - Simple file upload with special chars..."
$DU -q upload testData/file\ 2* du_tests
check_exit

#Checking with list
echo -ne " - Checking file..."
$DU -q list du_tests | grep "file 2 ù.txt" > /dev/null
check_exit

#Recursive directory upload
echo -ne " - Recursive directory upload..."
$DU -q upload testData/recurse du_tests
check_exit

#Recursive directory download
echo -ne " - Recursive directory download..."
$DU -q download du_tests/recurse
check_exit

#Checking the downloaded dir
echo -ne " - Checking the downloaded dir..."
diff -r recurse testData/recurse/
check_exit

#Again, recursive directory download
echo -ne " - Again recursive directory download..."
$DU -q download du_tests/recurse
check_exit

#Again, checking the downloaded dir
echo -ne " - Checking the downloaded dir..."
diff -r recurse testData/recurse/
check_exit

rm -fr "recurse"
rm -fr "testData"

#Rmdir
echo -ne " - Remove remote directory..."
$DU -q remove du_tests
check_exit


uninstall.sh
#!/usr/bin/env bash
#
# Dropbox Uploader and Dropbox Shell uninstaller
#
# Copyright (C) 2014 Stefanos Kalantzis <steve.tcmg@gmail.com>
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
#


# stored configuration options from installation
#installation_dir=
#installation_method=
#dropbox_uploader_command_name=
#dropbox_shell_command_name=
#downloaded_dir=

if [ -r ~/.dropbox_uploader_installed.conf ]; then
	echo -e "Reading used installation config options from ~/.dropbox_uploader_installed.conf\n"
	for line in $(cat ~/.dropbox_uploader_installed.conf); do
		eval $line
	done
else
	echo "~/.dropbox_uploader_installed.conf does not exist. Assuming it is not installed."
	exit
fi

echo -e "Removing installed files...\n"

eval rm -vf "$installation_dir/$dropbox_shell_command_name"
eval rm -vf "$installation_dir/$dropbox_uploader_command_name"
eval rm -vf "~/.dropbox_uploader_installed.conf"

if [ "$installation_method" == "link" ]; then
	echo -e "\nRestoring original files"
	eval mv -v "$downloaded_dir/dropShell.sh.orig" "$downloaded_dir/dropShell.sh"
	eval mv -v "$downloaded_dir/dropShell-fix-path.patch.orig" "$downloaded_dir/dropShell-fix-path.patch"
fi

echo -e "\n...done"



plugins/thunar/install.sh
#!/usr/bin/env bash

pack1=`dpkg -l |grep 'xclip' | wc -l`

if [[ $pack1 = 0 ]] ; then
    echo "Please install xclip package before we continue!"
else

DIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )"
script_name="thunar-dropbox.sh"
td_sh=$DIR/$script_name

rand1=$((RANDOM%1499999999999999+1400000000000000))
rand2=$((RANDOM%29+1))
rand=$rand1'-'$rand2

uca_xml="$HOME/.config/Thunar/uca.xml"

installed=`cat $uca_xml | grep thunar-dropbox.sh | wc -l`

if [[ $installed != 0 ]]; then
    echo "Plugin already installed!"
else
cat >thunarDropCustAction.txt <<EOL
<action>
    <icon>emblem-shared</icon>
    <name>Dropbox: share link</name>
    <unique-id>$rand</unique-id>
    <command>$td_sh %f</command>
    <description></description>
    <patterns>*</patterns>
    <directories/>
    <audio-files/>
    <image-files/>
    <other-files/>
    <text-files/>
    <video-files/>
</action>
EOL
    echo "Installing plugin..."
    
    if lno=$(sed -n -e '/<\/actions>/=' $uca_xml)
    then let lno=lno-1
         sed -i -e "$lno"'r thunarDropCustAction.txt' $uca_xml
    fi

    rm thunarDropCustAction.txt
    echo "Done!"
fi

fi


dropShell.sh
#!/usr/bin/env bash
#
# DropShell
#
# Copyright (C) 2013-2014 Andrea Fabrizi <andrea.fabrizi@gmail.com>
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
#

#Looking for dropbox uploader
if [ -f "./dropbox_uploader.sh" ]; then
    DU="./dropbox_uploader.sh"
else
    DU=$(which dropbox_uploader.sh)
    if [ $? -ne 0 ]; then
        echo "Dropbox Uploader not found!"
        exit 1
    fi
fi

#For MacOSX, install coreutils (which includes greadlink)
# $brew install coreutils
if [ "${OSTYPE:0:6}" == "darwin" -o "${OSTYPE:0:7}" == "freebsd" ]; then
    READLINK="greadlink"
else
    READLINK="readlink"
fi

SHELL_HISTORY=~/.dropshell_history
DU_OPT="-q"
BIN_DEPS="id $READLINK ls basename ls pwd cut"
VERSION="0.2"

umask 077

#Dependencies check
for i in $BIN_DEPS; do
    which $i > /dev/null
    if [ $? -ne 0 ]; then
        echo -e "Error: Required program could not be found: $i"
        exit 1
    fi
done

#Check DropBox Uploader
if [ ! -f "$DU" ]; then
    echo "Dropbox Uploader not found: $DU"
    echo "Please change the 'DU' variable according to the Dropbox Uploader location."
    exit 1
else
    DU=$($READLINK -m "$DU")
fi

#Returns the current user
function get_current_user
{
    id -nu
}

function normalize_path
{
    $READLINK -m "$1"
}

################
#### START  ####
################

echo -e "DropShell v$VERSION"
echo -e "The Interactive Dropbox SHELL"
echo -e "Andrea Fabrizi - andrea.fabrizi@gmail.com\n"
echo -e "Type help for the list of the available commands.\n"

history -r "$SHELL_HISTORY"
username=$(get_current_user)

#Initial Working Directory
CWD="/"

function sh_ls
{
    local arg1=$1

    #Listing current dir
    if [ -z "$arg1" ]; then
        "$DU" $DU_OPT list "$CWD"

    #Listing $arg1
    else

        #Relative or absolute path?
        if [ ${arg1:0:1} == "/" ]; then
            "$DU" $DU_OPT list "$(normalize_path "$arg1")"
        else
            "$DU" $DU_OPT list "$(normalize_path "$CWD/$arg1")"
        fi

        #Checking for errors
        if [ $? -ne 0 ]; then
            echo -e "ls: cannot access '$arg1': No such file or directory"
        fi
    fi
}

function sh_cd
{
    local arg1=$1

    OLD_CWD=$CWD

    if [ -z "$arg1" ]; then
        CWD="/"
    elif [ ${arg1:0:1} == "/" ]; then
        CWD=$arg1
    else
        CWD=$(normalize_path "$OLD_CWD/$arg1/")
    fi

    "$DU" $DU_OPT list "$CWD" > /dev/null

    #Checking for errors
    if [ $? -ne 0 ]; then
        echo -e "cd: $arg1: No such file or directory"
        CWD=$OLD_CWD
    fi
}

function sh_get
{
    local arg1=$1
    local arg2=$2

    if [ ! -z "$arg1" ]; then

        #Relative or absolute path?
        if [ ${arg1:0:1} == "/" ]; then
            "$DU" $DU_OPT download "$(normalize_path "$arg1")" "$arg2"
        else
            "$DU" $DU_OPT download "$(normalize_path "$CWD/$arg1")" "$arg2"
        fi

        #Checking for errors
        if [ $? -ne 0 ]; then
            echo -e "get: Download error"
        fi

    #args error
    else
        echo -e "get: missing operand"
        echo -e "syntax: get <FILE/DIR> [LOCAL_FILE/DIR]"
    fi
}

function sh_put
{
    local arg1=$1
    local arg2=$2

    if [ ! -z "$arg1" ]; then

        #Relative or absolute path?
        if [ "${arg2:0:1}" == "/" ]; then
            "$DU" $DU_OPT upload "$arg1" "$(normalize_path "$arg2")"
        else
            "$DU" $DU_OPT upload "$arg1" "$(normalize_path "$CWD/$arg2")"
        fi

        #Checking for errors
        if [ $? -ne 0 ]; then
            echo -e "put: Upload error"
        fi

    #args error
    else
        echo -e "put: missing operand"
        echo -e "syntax: put <FILE/DIR> <REMOTE_FILE/DIR>"
    fi
}

function sh_rm
{
    local arg1=$1

    if [ ! -z "$arg1" ]; then

        #Relative or absolute path?
        if [ ${arg1:0:1} == "/" ]; then
            "$DU" $DU_OPT remove "$(normalize_path "$arg1")"
        else
            "$DU" $DU_OPT remove "$(normalize_path "$CWD/$arg1")"
        fi

        #Checking for errors
        if [ $? -ne 0 ]; then
            echo -e "rm: cannot remove '$arg1'"
        fi

    #args error
    else
        echo -e "rm: missing operand"
        echo -e "syntax: rm <FILE/DIR>"
    fi
}

function sh_mkdir
{
    local arg1=$1

    if [ ! -z "$arg1" ]; then

        #Relative or absolute path?
        if [ ${arg1:0:1} == "/" ]; then
            "$DU" $DU_OPT mkdir "$(normalize_path "$arg1")"
        else
            "$DU" $DU_OPT mkdir "$(normalize_path "$CWD/$arg1")"
        fi

        #Checking for errors
        if [ $? -ne 0 ]; then
            echo -e "mkdir: cannot create directory '$arg1'"
        fi

    #args error
    else
        echo -e "mkdir: missing operand"
        echo -e "syntax: mkdir <DIR_NAME>"
    fi
}

function sh_mv
{
    local arg1=$1
    local arg2=$2

    if [ ! -z "$arg1" -a ! -z "$arg2" ]; then

        #SRC relative or absolute path?
        if [ ${arg1:0:1} == "/" ]; then
            SRC="$arg1"
        else
            SRC="$CWD/$arg1"
        fi

        #DST relative or absolute path?
        if [ ${arg2:0:1} == "/" ]; then
            DST="$arg2"
        else
            DST="$CWD/$arg2"
        fi

        "$DU" $DU_OPT move "$(normalize_path "$SRC")" "$(normalize_path "$DST")"

        #Checking for errors
        if [ $? -ne 0 ]; then
            echo -e "mv: cannot move '$arg1' to '$arg2'"
        fi

    #args error
    else
        echo -e "mv: missing operand"
        echo -e "syntax: mv <FILE/DIR> <DEST_FILE/DIR>"
    fi
}

function sh_cp
{
    local arg1=$1
    local arg2=$2

    if [ ! -z "$arg1" -a ! -z "$arg2" ]; then

        #SRC relative or absolute path?
        if [ ${arg1:0:1} == "/" ]; then
            SRC="$arg1"
        else
            SRC="$CWD/$arg1"
        fi

        #DST relative or absolute path?
        if [ ${arg2:0:1} == "/" ]; then
            DST="$arg2"
        else
            DST="$CWD/$arg2"
        fi

        "$DU" $DU_OPT copy "$(normalize_path "$SRC")" "$(normalize_path "$DST")"

        #Checking for errors
        if [ $? -ne 0 ]; then
            echo -e "cp: cannot copy '$arg1' to '$arg2'"
        fi

    #args error
    else
        echo -e "cp: missing operand"
        echo -e "syntax: cp <FILE/DIR> <DEST_FILE/DIR>"
    fi
}

function sh_free
{
    "$DU" $DU_OPT info | grep "Free:" | cut -f 2
}

function sh_cat
{
    local arg1=$1

    if [ ! -z "$arg1" ]; then

        tmp_cat="/tmp/sh_cat_$RANDOM"
        sh_get "$arg1" "$tmp_cat"
        cat "$tmp_cat"
        rm -fr "$tmp_cat"

    #args error
    else
        echo -e "cat: missing operand"
        echo -e "syntax: cat <FILE>"
    fi
}

while (true); do

    #Reading command from shell
    read -e -p "$username@Dropbox:$CWD$ " input

    #Tokenizing command
    eval tokens=($input)
    cmd=${tokens[0]}
    arg1=${tokens[1]}
    arg2=${tokens[2]}

    #Saving command in the history file
    history -s "$input"
    history -w "$SHELL_HISTORY"

    case $cmd in

        ls)
            sh_ls "$arg1"
        ;;

        cd)
            sh_cd "$arg1"
        ;;

        pwd)
            echo $CWD
        ;;

        get)
            sh_get "$arg1" "$arg2"
        ;;

        put)
            sh_put "$arg1" "$arg2"
        ;;

        rm)
            sh_rm "$arg1"
        ;;

        mkdir)
            sh_mkdir "$arg1"
        ;;

        mv)
            sh_mv "$arg1" "$arg2"
        ;;

        cp)
            sh_cp "$arg1" "$arg2"
        ;;

        cat)
            sh_cat "$arg1"
        ;;

        free)
            sh_free
        ;;

        lls)
            ls -l
        ;;

        lpwd)
            pwd
        ;;

        lcd)
            cd "$arg1"
        ;;

        help)
            echo -e "Supported commands: ls, cd, pwd, get, put, cat, rm, mkdir, mv, cp, free, lls, lpwd, lcd, help, exit\n"
        ;;

        quit|exit)
            exit 0
        ;;

        *)
            if [ ! -z "$cmd" ]; then
                echo -ne "Unknown command: $cmd\n"
            fi
        ;;
    esac
done
